### PR TITLE
core/libraries: Videodec2 implementation

### PR DESCRIPTION
**Pull request by polybiusproxy**
Work-in-progress implementation of the Videodec2 library, used for AVC (H.264) and HEVC (H.265) video decoding.
Seen on games such as Persona 5 Royal and Rock Band 4.

Left tasks:
- [x] Error handling
- [x] Decoder flushing
- [x] Decoder reset
- [x] Obtaining picture info
- [ ] HEVC/H.265 implementation
- [ ] Implement memory querying

The memory querying functions are quite important, as the sizes we output can affect games - for example, if we give a low size for memory querying on Persona 5 Royal, the decode output breaks (see the first image below).
The decoder library outputs a different size depending on the decoder profile given by the game.

<p align="center">
  
  
  
  
</p>

### CODE REVIEW

**Comment by polybiusproxy** [comment]
Fixes crashes on Persona 5 Royal.
imagen

**Comment by georgemoralis** [comment]
@vladmikhalin can you review this?

**Comment by roamic** [comment]
> @vladmikhalin can you review this?

It's still a draft, is it ready for review?

**Comment by georgemoralis** [comment]
it is ready

**Comment by polybiusproxy** [comment]
> > @vladmikhalin can you review this?
> 
> It's still a draft, is it ready for review?

It's still marked as a draft but I'd still want you to review it

File: src/core/libraries/videodec/videodec2_impl.cpp [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    packet->size = inputData.auSize;
+
+    int ret = avcodec_send_packet(mCodecContext, packet);
+    if (ret < 0) {
+        av_packet_free(&packet);
+        return 0;
+    }
+
+    AVFrame* frame = av_frame_alloc();
+    ret = avcodec_receive_frame(mCodecContext, frame);
```
Comment (roamic) [inline src/core/libraries/videodec/videodec2_impl.cpp]: One packet can contain several frames. You should drain them all until `AVERROR(EAGAIN)` or `AVERROR_EOF` is returned. What you can do here is call `avcodec_receive_frame` before `avcodec_send_packet` to query the decoder on available frames. Only if it returns `AVERROR(EAGAIN)` you should call `avcodec_send_packet` again.

File: src/core/libraries/videodec/videodec2_impl.cpp [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+                      SceVideodec2OutputInfo& outputInfo) {
+    AVPacket* packet = av_packet_alloc();
+
+    frameBuffer.isAccepted = false;
+    outputInfo.isValid = false;
+    outputInfo.isErrorFrame = true;
+    outputInfo.pictureCount = 0;
+
+    packet->data = (u8*)inputData.pAuData;
+    packet->size = inputData.auSize;
```
Comment (roamic) [inline src/core/libraries/videodec/videodec2_impl.cpp:65]: You can set `packet->pts` to `inputData.ptsData` and `packet->dts` to `inputData.dtsData`

**Review by roamic — COMMENTED** [review]
Apart from the obvious magic numbers and bogus values that I can attribute to this PR being a draft I only have these 2 comments.

**Comment by georgemoralis** [comment]
what other games uses videcodec2 expect persona????

**Comment by polybiusproxy** [comment]
Rock Band 4 and lots of Unreal Engine games, it seems

**Comment by polybiusproxy** [comment]
I believe this is ready to merge. The remaining tasks can be left for later.

File: src/core/libraries/videodec/videodec2.cpp [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+
+    if (!pOutputInfoIn) {
+        return ORBIS_VIDEODEC2_ERROR_ARGUMENT_POINTER;
+    }
+    if (pOutputInfoIn->thisSize != sizeof(OrbisVideodec2OutputInfo)) {
+        return ORBIS_VIDEODEC2_ERROR_STRUCT_SIZE;
+    }
+
+    if (pOutputInfoIn->pictureCount == 0 || gPictureInfos.empty()) {
+        return 0;
```
Comment (roamic) [inline src/core/libraries/videodec/videodec2.cpp]:
```suggestion
        return ORBIS_OK;

```

File: src/core/libraries/videodec/videodec2_impl.cpp [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+            pictureInfo.frameCropTopOffset = frame->crop_top;
+            pictureInfo.frameCropBottomOffset = frame->crop_bottom;
+
+            gPictureInfos.push_back(pictureInfo);
+        }
+    }
+
+    av_packet_free(&packet);
+    av_frame_free(&frame);
+    return 0;
```
Comment (roamic) [inline src/core/libraries/videodec/videodec2_impl.cpp]:
```suggestion
    return ORBIS_OK;

```

File: src/core/libraries/videodec/videodec2_impl.cpp [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+
+        outputInfo.isValid = true;
+        outputInfo.isErrorFrame = false;
+        outputInfo.pictureCount++;
+
+        // FIXME: Should we add picture info here too?
+    }
+
+    av_frame_free(&frame);
+    return 0;
```
Comment (roamic) [inline src/core/libraries/videodec/videodec2_impl.cpp]:
```suggestion
    return ORBIS_OK;

```

File: src/core/libraries/videodec/videodec2_impl.cpp [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+}
+
+s32 VdecDecoder::Reset() {
+    avcodec_flush_buffers(mCodecContext);
+    gPictureInfos.clear();
+    return 0;
```
Comment (roamic) [inline src/core/libraries/videodec/videodec2_impl.cpp]:
```suggestion
    return ORBIS_OK;

```

**Review by roamic — APPROVED** [review]
LGTM otherwise

File: src/core/libraries/videodec/videodec2.cpp [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+    pMemInfoOut->pCpuMemory = nullptr;
+    pMemInfoOut->pGpuMemory = nullptr;
+    pMemInfoOut->pCpuGpuMemory = nullptr;
+
+    pMemInfoOut->cpuGpuMemorySize = kMinimumMemorySize;
+    pMemInfoOut->cpuMemorySize = kMinimumMemorySize;
+    pMemInfoOut->gpuMemorySize = kMinimumMemorySize;
+
+    pMemInfoOut->maxFrameBufferSize = kMinimumMemorySize;
+    pMemInfoOut->frameBufferAlignment = kMinimumMemorySize;
```
Comment (roamic) [inline src/core/libraries/videodec/videodec2.cpp]: Oh, shouldn't this be 0x100 or something?

Comment (polybiusproxy) [reply src/core/libraries/videodec/videodec2.cpp]: I'm unsure, no clue on what should these values be.

Comment (roamic) [reply src/core/libraries/videodec/videodec2.cpp]: Try 0x100 and see if it fixes/breaks anything. This value is used to align framebuffer data for AvPlayer.

Comment (polybiusproxy) [reply src/core/libraries/videodec/videodec2.cpp]: What do you think should the max framebuffer size be?

Comment (roamic) [reply src/core/libraries/videodec/videodec2.cpp]: I think you can leave it as is

Comment (polybiusproxy) [reply src/core/libraries/videodec/videodec2.cpp]: Got it, I'll try this later.

Comment (polybiusproxy) [reply src/core/libraries/videodec/videodec2.cpp]: > Try 0x100 and see if it fixes/breaks anything. This value is used to align framebuffer data for AvPlayer.

No change whatsoever.

**Pull request merged**